openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -900,6 +907,102 @@ impl<T, B: HardwareIsolatedBacking> hv1_hypercall::ModifyVtlProtectionMask
     }
 }
 
+impl<T, B: HardwareIsolatedBacking> hv1_hypercall::TranslateVirtualAddressX64
+    for UhHypercallHandler<'_, '_, T, B>
+{
+    fn translate_virtual_address(
+        &mut self,
+        partition_id: u64,
+        vp_index: u32,
+        control_flags: hvdef::hypercall::TranslateGvaControlFlagsX64,
+        gva_page: u64,
+    ) -> HvResult<hvdef::hypercall::TranslateVirtualAddressOutput> {
+        if partition_id != hvdef::HV_PARTITION_ID_SELF {
+            return Err(HvError::AccessDenied);
+        }
+
+        if vp_index != hvdef::HV_VP_INDEX_SELF && vp_index != self.vp.vp_index().index() {
+            return Err(HvError::AccessDenied);
+        }
+
+        let target_vtl = self
+            .target_vtl_no_higher(
+                control_flags
+                    .input_vtl()
+                    .target_vtl()?
+                    .unwrap_or(self.intercepted_vtl.into()),
+            )
+            .map_err(|_| HvError::AccessDenied)?;
+
+        if self.intercepted_vtl == target_vtl {
+            return Err(HvError::AccessDenied);
+        }
+
+        let gva = gva_page * hvdef::HV_PAGE_SIZE;
+
+        let registers = self.vp.backing.translation_registers(self.vp, target_vtl);
+
+        let pat = self
+            .vp
+            .access_state(target_vtl.into())
+            .cache_control()
+            .map_err(Self::reg_access_error_to_hv_err)?
+            .msr_cr_pat;
+
+        if control_flags.tlb_flush_inhibit() {
+            self.vp
+                .set_tlb_lock(self.intercepted_vtl.into(), target_vtl);
+        }
+
+        match virt_support_x86emu::translate::translate_gva_to_gpa(
+            &self.vp.partition.gm[target_vtl], // TODO GUEST VSM: This doesn't have VTL access checks.
+            gva,
+            &registers,
+            virt_support_x86emu::translate::TranslateFlags::from_hv_flags(control_flags),
+        ) {
+            Ok(virt_support_x86emu::translate::TranslateResult { gpa, cache_info }) => {
+                let overlay_page = hvdef::hypercall::MsrHypercallContents::from(
```

Comment:
I believe this is not the only overlay page, is it? 

We really need to come up with a consistent model for overlay page handling. I'd suggest skipping this handling for now and add a TODO + open an issue, unless we actually care about this handling with high priority (do guests really use this overlay page bit in the translate result in practice? What do they do with it?)

Reply:
Currently the guest is using it to validate that the "overlay state" of the page is as it expects. And it's only attempting this on pages it thinks are part of a driver or associated with executable code. The hypercall code page fits that description, and it seems possible to me that we could trigger some hyperguard failures without it. So I'm inclined to leave the hypercall code page here, but still put a TODO + open an issue for more general overlay page support, if that makes sense to you.

Reply:
OK, that's fine.